### DISCUSSION TITLE
Scenario: Online Video Streaming Platform (Bounded Contexts, Aggregates, Domain and Application Services)

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/iSAQB-CPSA-A-DDD#6] [Hands-on]
The scenario and tasks are tailored to bring out the differences between bounded contexts and aggregates, as well as domain services and application services, in a way that is directly applicable to real architectural decision-making.

---

## Scenario: Online Video Streaming Platform

Your organization is building an online video streaming service with multiple **bounded contexts**:

1. **Content Catalog Context**:  
   - Manages information about TV shows, movies, and episodes.
   - Typical Aggregates:  
     - **Show**: Aggregate root representing a series, containing episodes as entities or referencing them.  
     - **Episode**: Potentially its own aggregate if episodes have distinct lifecycles or as part of the Show aggregate.

2. **Subscription Billing Context**:  
   - Manages user subscriptions, invoicing, and payments.
   - Typical Aggregates:  
     - **Subscription**: Aggregate root representing a user’s subscription plan and status.  
     - **Invoice**: Aggregate root representing billing events tied to a subscription.

3. **User Profile Context**:  
   - Manages user profiles, preferences, watchlists, and viewing history.
   - Typical Aggregates:  
     - **UserProfile**: Aggregate root for storing preferences, watchlists.
   
Users interact with the system to browse shows, manage subscriptions, and watch episodes. Each bounded context has its own domain model, and your task is to identify where domain and application services fit, as well as apply rules of thumb to distinguish bounded contexts from aggregates.

---

## Rules of Thumb (To Be Applied in the Tasks)

1. **Bounded Context vs. Aggregate**:  
   - A **bounded context** sets a clear boundary where a specific model and ubiquitous language apply. It often aligns with a subdomain of the business (e.g., "Subscription Billing").
   - An **aggregate** is a consistency boundary inside a bounded context. It ensures that all changes within it are consistent and transactional. Different aggregates in the same bounded context represent different aspects of the domain model.
   - **Apply This**: Ensure you do not confuse a large bounded context with a single aggregate. A context can have multiple aggregates. The aggregate addresses transactional invariants; the bounded context addresses conceptual and linguistic boundaries.

2. **Domain Service vs. Application Service**:  
   - **Domain Service**: Encapsulates complex domain logic that doesn’t fit into a single entity or aggregate and operates purely on domain concepts. It has no infrastructural concerns, no user interface logic, and no orchestration logic.  
   - **Application Service**: Implements use-cases by orchestrating calls to domain services, repositories, and aggregates. It handles input/output from the outside world and coordinates steps, but does not contain core domain logic itself.
   - **Apply This**: When you see logic that combines data from multiple aggregates to compute a domain decision (e.g., applying a discount to a subscription based on viewing habits), think: Domain Service. When you see a workflow ("User signs up, system creates subscription, sends welcome email"), that’s typically Application Service logic.

3. **Heuristic for Identifying Domain Services**:  
   - If the logic:  
     - Does not depend on user interaction directly (just inputs/outputs domain concepts),  
     - Involves multiple domain objects from the same or different aggregates,  
     - And expresses a pure business rule,  
     Then it’s likely a Domain Service.

4. **Heuristic for Identifying Application Services**:  
   - If the logic:  
     - Describes a sequence of steps to fulfill a use-case, involving calling multiple repositories or domain services,  
     - Handles authentication/authorization or triggers events to other bounded contexts,  
     Then it’s likely an Application Service.

---

### Task 1: Distinguish Bounded Contexts from Aggregates

**Situation:** In the Content Catalog Context, you have a `Show` aggregate. Students often confuse the entire Content Catalog as one big aggregate.

**Task:**  
Identify at least two aggregates in the Content Catalog Context and explain why they are aggregates rather than separate bounded contexts. Consider if `Episode` is part of the `Show` aggregate or a separate aggregate. Apply the rule of thumb:  
- **Write down in one paragraph**: Which aggregates you see in the Content Catalog (e.g., `Show` as root, `Episode` as part of `Show` or separate?), and why the entire Content Catalog is not a single aggregate but a bounded context containing multiple aggregates.

### Task 2: Applying Domain vs. Application Service Rules

**Situation:** The user wants to watch an episode. The use-case steps might be:  
1. Verify user subscription is active (check Subscription aggregate in Subscription Billing Context).  
2. Verify episode availability (check Show or Episode aggregate in Content Catalog Context).  
3. If subscription is valid and episode is available, record "episode watched" in UserProfile aggregate (User Profile Context).

**Task:**  
Break down the scenario into what would be handled by an Application Service and what might be handled by a Domain Service. Consider the following logic:  
- If a user watches more than 3 episodes in a day, apply a promotional credit to their subscription next month. This involves checking user’s viewing count (User Profile Context) and updating their subscription discount (Subscription Billing Context).  
- According to the rules of thumb, the logic that determines whether a discount applies is pure domain logic (Domain Service), while the orchestration of checking user’s profile, updating subscription, and maybe notifying the user is Application Service logic.

**Your Task Steps:**
1. **Identify a Domain Service**:  
   Which part of the scenario is pure domain logic? Write a short description of a `ViewingBonusDomainService` that takes user’s daily watch count and returns a discount decision. This service does not send emails or query external systems; it only implements the business rule: “>3 episodes = discount next month.”
   
2. **Identify an Application Service**:  
   Write a short outline of an `EpisodeWatchingApplicationService` that:  
   - Takes a "WatchEpisode" request (UserID, EpisodeID).
   - Calls repositories to load `UserProfile` and `Show` aggregates.
   - Uses the `ViewingBonusDomainService` to determine if a discount applies.
   - Updates `Subscription` aggregate accordingly.
   - This outline should show that the Application Service orchestrates multiple steps, but does not contain the discount logic itself.

### Task 3: Reflect on Bounded Context Boundaries

**Situation:** The Subscription Billing Context and Content Catalog Context are different bounded contexts. Students sometimes mix them up.

**Task:**  
Write a short explanation (2-3 sentences) on why `Subscription` belongs to the Subscription Billing Context and not the Content Catalog Context. Use the rule of thumb: Bounded contexts split the domain into conceptual areas with their own ubiquitous language. Illustrate how `Subscription` and `Show` use different terminologies and solve different problems, thus belong to different bounded contexts.

### Task 4: Additional CPSA-Level Architect Activities

**Situation:** As an architect, you must also consider quality attributes, integration patterns, and compliance.

**Task:**  
- Propose an integration approach for when the Application Service in the User Profile Context needs to update the Subscription in the Subscription Billing Context (e.g., via domain events or synchronous calls?).  
- State a quality attribute (e.g., scalability) and how separating bounded contexts helps.  
- Mention a compliance or security consideration (e.g., payment details protected in the Subscription Billing Context) and how a domain service that calculates discounts doesn’t need direct access to payment info, thus reducing exposure.

**Here, you apply the rules of thumb for architectural decisions:**  
- Keep sensitive data in the context that needs it.
- Use asynchronous domain events if eventual consistency is acceptable.
- Keep domain services lean and stateless, focusing on rules, not infrastructure.

---

## Desired Outcomes

By doing these tasks:

- Students will **clearly separate** the concept of bounded contexts (high-level domain partitions) from aggregates (transactional consistency boundaries within a context).
- They will produce concrete examples of **domain services** (pure business logic spanning multiple aggregates) and **application services** (orchestration of steps, I/O handling).
- They will practice **architectural decision-making** at a CPSA-A level, considering integration, quality attributes, and compliance.
- They will have concrete outputs: written paragraphs, short outlines of services, and rationales for their decisions, reinforcing the rules of thumb given.

---

## Expert Suggestions to Students

1. **When Unsure, Ask ‘What Layer?’**: If your code decides how to compute a discount purely based on domain rules and domain objects—Domain Service. If it calls multiple repositories, interacts with external systems, or sequences steps of a use-case—Application Service.
2. **Bounded Context First, Aggregate Second**: Always determine bounded contexts (broad conceptual boundaries) before drilling down to aggregates. Aggregates live inside bounded contexts, not the other way around.
3. **Check for Business Logic Location**: If you find business rules scattered in your application services, move them to domain services or aggregates. Application services should be thin coordinators.
4. **Document Your Decisions**: For each choice (Domain vs. Application Service, Bound Context splits), write down why. In an exam or real project, showing your reasoning is as important as the final model.

**Comment by akash-coded** [comment]
## Solution

---

## Key Concepts Recap

- **Bounded Context**: A large logical boundary within the domain, where a specific, consistent model and language are used. Think of it as a "domain area" (e.g., Content Catalog, Subscription Billing, User Profile).

- **Aggregate**: A smaller, transactional consistency boundary inside a bounded context. An aggregate ensures that certain business rules (invariants) are always true whenever you modify its data.

- **Domain Service**: A service that contains **pure business logic**. It doesn’t handle input from users or coordinate multiple steps of a process. It just applies domain rules, often using data from multiple aggregates.

- **Application Service**: A service that implements a **use-case or workflow**. It may call multiple domain services, load and save aggregates from repositories, handle input/output, and orchestrate the overall process, but it does not contain the domain logic itself.

---

## Scenario (Online Video Streaming Platform)

We have three bounded contexts:

1. **Content Catalog Context**:  
   Focuses on what shows and episodes exist and their availability.  
   - Aggregates might include:  
     **Show (Aggregate Root)**: Contains information about the TV show (title, description) and possibly a list of episodes.  
     **Episode**: Could be part of the Show aggregate or, if episodes have complex lifecycles, another aggregate. Usually, a Show and its Episodes belong to a single aggregate (the Show), ensuring consistency of the show’s content.

2. **Subscription Billing Context**:  
   Focuses on user subscriptions, plans, and billing.  
   - Aggregates might include:  
     **Subscription (Aggregate Root)**: Holds the user's subscription details (plan type, start date, status).  
     **Invoice (Aggregate Root)**: Represents a billing event or charge.

3. **User Profile Context**:  
   Focuses on user preferences, watch history, and watchlists.  
   - Aggregates might include:  
     **UserProfile (Aggregate Root)**: Stores the user’s watch history, preferences, and lists of favorite shows.

**Important:**  
- Each bounded context (Content Catalog, Subscription Billing, User Profile) is its own domain "area."  
- Each context contains one or more aggregates. For example, the Subscription Billing Context has a `Subscription` aggregate and an `Invoice` aggregate.

---

## Distinguishing Bounded Contexts from Aggregates

- **Bounded Context** is larger: For example, the **Content Catalog Context** is one big area that deals with shows and episodes.  
- **Aggregate** is smaller and exists inside a single bounded context: For example, the **Show** aggregate is inside the Content Catalog context. It ensures that all changes to a show (like adding an episode) remain consistent.

**Simple Rule of Thumb**:  
- Bounded Context = Big box (e.g., “Subscription Billing” as a theme).  
- Aggregates = Smaller boxes inside it (e.g., “Subscription” as one box, “Invoice” as another, both inside “Subscription Billing”).

---

## Domain Service vs. Application Service

**Domain Service**: Pure logic about business rules.  
- Example: You have a rule that says, “If the user watches more than 3 episodes in a day, give them a discount on their next subscription invoice.”  
- This logic doesn’t care about how data is fetched or how to inform the user. It just calculates whether a discount applies or not based on inputs (e.g., number of episodes watched, current subscription status).

**Application Service**: Orchestrates a process or a use-case.  
- Example: When the user presses “Watch Episode,” the application service:  
  1. Checks if the subscription is active (by loading the `Subscription` aggregate from the Subscription Billing context).  
  2. Checks if the episode is available (by querying the `Show` aggregate in the Content Catalog context).  
  3. Updates the `UserProfile` aggregate to record that the user watched the episode.  
  4. Calls the domain service (e.g., `ViewingBonusDomainService`) to determine if a discount should be applied.  
  5. If a discount is due, updates the `Subscription` aggregate accordingly.  
  6. Returns a response to the user.

**Simple Rule of Thumb**:  
- If it’s orchestrating steps, talking to multiple repositories, or handling user input/output → Application Service.  
- If it’s just pure domain logic, like “given these facts, what’s the discount?” → Domain Service.

---

## Example Application of the Rules

**Scenario**: User watches an episode.

- **Domain Service** Example: `ViewingBonusDomainService`  
  Input: User’s daily watch count, subscription level  
  Output: Discount percentage (if any)  
  This service does not send emails or save data. It only applies the business rule: “more than 3 episodes = discount.”

- **Application Service** Example: `WatchEpisodeApplicationService`  
  Steps:  
  1. Receive request (UserID, EpisodeID).  
  2. Load `UserProfile` aggregate to increment watch count.  
  3. Load `Show` to check episode availability.  
  4. Load `Subscription` to ensure user is subscribed.  
  5. Call `ViewingBonusDomainService` to check discount.  
  6. If discount applies, update `Subscription` aggregate.  
  7. Return a success message to the UI.

Notice: The Application Service coordinates everything; the Domain Service just applies the rule.

---

## Bounded Context Boundaries

- The `Subscription` aggregate belongs to the Subscription Billing Context because it deals with plans and payments.  
- The `Show` aggregate belongs to the Content Catalog Context because it deals with the catalog of episodes.  
- Different contexts solve different problems with their own models and terms. The Subscription Billing Context talks about “plans” and “invoices,” while the Content Catalog Context talks about “shows” and “episodes.”

This ensures that each context can evolve independently and is easier to maintain. Aggregates within each context ensure transactional consistency on a smaller scale.

---

## Quality and Architectural Decisions

- By separating logic into bounded contexts, you can scale them independently. For example, the Content Catalog might need more caching due to heavy reads, while Subscription Billing might need strong consistency for financial transactions.
- By using domain services for logic and application services for orchestration, your code stays cleaner. Domain rules don’t get tangled up with workflow steps or external integration concerns.
- This approach also makes it easier to secure sensitive data, like payment info in the Subscription Billing Context, and keep it separate from the Content Catalog Context.

---

## Conclusion

- **Bounded contexts**: Big conceptual boundaries grouping related domain concepts.  
- **Aggregates**: Smaller consistency boundaries inside a bounded context.  
- **Domain services**: Pure business logic that doesn’t belong to a single entity.  
- **Application services**: Orchestrate use-cases, handle input/output, and coordinate calls to domain services and aggregates.

**Comment by shruti-kirti-vwits** [comment]
1. Content Catalog Context:

Domain(Bounded Context) : Content Catalog 
Aggregate: Show(Episode), Movies, Live Steam

Explanation:
The episode is tractional with Show, hence part of the same aggregate.
Movies and LiveStreams are individualistic and are not transactions with Show or each other; hence, they have separate aggregates with their own entities: Movie and Steam, respectively.


2. Domain service vs Application Service:
Domain Service:
ViewingBonusDomainService: It checks user's watchcount >=3 and enables Promotional Code for next month.


Application Service:
EpisodeWatchingApplicationService: 
1. Calls UserProfile Service to Validate the subscription Status.
2. Calls EpisodeAvailaibity Service to identify if the Episode requested exist or not. 
3. ContentAvailbility Service to Validate the content is available for the region accesses.
4. Start the PlayEpisode Service to stream
5. Call WatchList Service to store the count of epiode watched.


3.
Subscription:
Subscription has properties like: Type, Payement Method
Behaviour: Bill Calculation, Auto Payment

Show:
A show has properties like : Genere, Language. 
It has behaviour like : Play, Download, Seek

As both of the Context deal with completely different entity. One has properties for the shows and other have all delatils wrt billing, hence they need to be in diffrent context.


4.
a. Domain Event: Billing cycle reached
The When for the user ID the cycle has come to end, The event is triggered and the subscription context is called via GenerateBill. The GenerateBill will call Domain Service - CalculateBill and then Call AutoDebit Application Service to connect with 3rd party and get money deducted.
Upon confirmation the User Context is called via UpdateUser Service and Set next Date for bill deduction.

b. Scalability: 
Adding the Option of UPI in the Subscription enables this to get added without problem.
Adding a new type of content can directly be added in content - example interactive games.

c. The Application service like GenrateBill will call a Domain service like CalucateBill and create the Bill for the payement.
